### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by okjake**
This PR will contain a review of all the room booker code

### CODE REVIEW

File: client/src/components/DayView/Calendar.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  resourcesFunc = ({ start }, successCallback, failureCallback) => {
+    const { rooms } = this.state;
+    successCallback(rooms.map((room) => ({ id: room.id, title: room.name })));
+
+    this.fetchRoomEvent(moment(start).format('YYYY-MM-DD')).catch(
+      failureCallback
+    );
+  };
+
+  handleDateSelect = ({ resource: { id: roomId }, start, end }) => {
```
Comment (okjake) [inline client/src/components/DayView/Calendar.js:81]: This is completely personal, but I find this two level destructuring hard to read

File: client/src/components/Form/AddRoom/index.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  if (updateID > 0) {
+    form.setFieldsValue({
+      name: initialName,
+    });
+  }
+
+  return (
+    <div>
+      <Button type="primary" onClick={onClick}>
+        New Room
```
Comment (okjake) [inline client/src/components/Form/AddRoom/index.js:24]: Not necessary here but something to think about - if you store user displayable strings like these in external files and import them, it will help a lot if you decide to internationalize the project later, you'll find it much easier to swap translated strings

File: client/src/components/Form/AddRoom/index.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      </Button>
+      <Modal
+        visible={visible}
+        title="Create a new room"
+        okText={updateID > 0 ? 'Update' : 'Create'}
+        cancelText="Cancel"
+        onCancel={onCancel}
+        onOk={() => {
```
Comment (okjake) [inline client/src/components/Form/AddRoom/index.js:32]: Try to avoid deep nesting inside components, it makes the code simpler to read if you just pass a function name here, and you'll find it easier to write unit tests. There are two patterns (that I'm aware of) that you can use to pass the context you need to need to an external function - either use a closure, a function that returns a function, for example:

```
const createOkHandler = (form, onUpdate, onCreate) => () => {
  // Note that this is a function that returns a function, the outer fn provides the context, the inner fn is the OK handler itself
  form.validateFields()... 
}
```
and use it in your Modal
```
<Modal onOk={createOkHandler(form, onUpdate, onCreate)}... />
```


or, alternatively, define a normal handler function:

```
const onOk = (form, onUpdate, onCancel) => {
  // Normal handler
  form.validateValues()...
}
```

and add context with `.bind`

```
<Modal onOk={onOk.bind(null, form, onUpdate, onCancel)} ... />
```

File: client/src/components/Form/Reservation/index.js [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+  Modal,
+  Radio,
+  Switch,
+  TimePicker,
+} from 'antd';
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+class BookingForm extends React.Component {
```
Comment (okjake) [inline client/src/components/Form/Reservation/index.js:16]: This component has got pretty big, I would start to look at how to split this up into several files.
Moving API calls into a helper library, and moving business logic into dedicated modules - things like findRoomIdByName will probably end up being useful in lots of places in the app, so it's good to define them as pure functions in separate files, then you can unit test them easily.

File: client/src/components/Form/Reservation/index.js [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  findRoomNameById = (id) => {
+    const { rooms } = this.props;
+    return rooms.find((room) => room.id === Number(id)).name;
+  };
+
+  repeatOnChange = (e) => {
+    this.setState({ repeat: e.target.value });
+  };
+
+  makeBookingArr = (
```
Comment (okjake) [inline client/src/components/Form/Reservation/index.js:74]: This is a good candidate to pull out into a separate file and unit test heavily, it will help you when weird bugs come up which will almost certainly happen (date logic is usually problematic ;)

File: server/__test__/booking.test.js [hunk 6]
```diff
@@ -0,0 +1,293 @@
+/* eslint-disable no-undef */
```
Comment (okjake) [inline server/__test__/booking.test.js:1]: These API tests are great

File: server/__test__/booking.test.js [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+const request = require('supertest');
+const app = require('../src/app');
+const dbBuild = require('../src/database/config/build');
+const connection = require('../src/database/config/connection.js');
+
+beforeEach(() => dbBuild());
+
+test('make new booking no overlapping', (done) => {
+  request(app)
```
Comment (okjake) [inline server/__test__/booking.test.js:10]: Totally preference but I'd probably make a very light abstraction for setting the content type and cookie here to save some typing, and allow us to change the cookie across all the tests easily later if necessary

File: server/__test__/booking.test.js [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    )
+    .expect(201)
+    .end((err) => {
+      if (err) return done(err);
+      return done();
+    });
+});
+
+test('make new booking no overlapping between 2021-04-14 14:30:00-2021-04-14 16:00:00 ', (done) => {
+  request(app)
```
Comment (okjake) [inline server/__test__/booking.test.js:40]: You can also use promises with supertest, if you return the result of `request` you don't need to worry about the done callback

Comment (okjake) [reply server/__test__/booking.test.js:40]: Again it's preference, but I would lean towards using Promises / async/await where possible because it's the direction JS development is moving in generally

File: server/__test__/index.test.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+/* eslint-disable no-undef */
+test('test of jest', () => {
+  expect(1).toBe(1);
```
Comment (okjake) [inline server/__test__/index.test.js:3]: :D I like this test

File: server/__test__/router.test.js [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+      'Content-Type': 'application/json',
+    })
+    .set('Cookie', [
+      'token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySUQiOjIsInJvbGUiOnRydWUsImlhdCI6MTU4NTgxNTc1MX0.SpdrsYcfCym_CIgCM4nocmHMULnF0yVx2DzkoMRFFqM',
+    ])
+    .send(JSON.stringify({ active: true }))
+    .expect(200)
+    .end((err) => {
+      if (err) return done(err);
+      getUserById(4).then(({ rows }) => expect(rows[0].is_active).toBe(true));
```
Comment (okjake) [inline server/__test__/router.test.js:276]: Be careful here, combining promises and callbacks is dangerous - I don't think this will do what you expect. getUserById is asynchronous so the function you pass to `.then` will be triggered after you've already called `done`

File: server/src/controllers/activateAccount.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  activateSchema
+    .validateAsync({ active, admin })
+    .catch((error) => {
+      throw Boom.badRequest(error.message);
+    })
+    .then(() => {
+      const promiseArray = [];
+      if (active !== undefined) promiseArray.push(activateUser(id, active));
+      if (admin !== undefined) promiseArray.push(makeAdmin(id, admin));
+      return Promise.all(promiseArray);
```
Comment (okjake) [inline server/src/controllers/activateAccount.js:24]: This will throw if any of the promises fail. What should happen if makeAdmin succeeded but activateUser failed for example?

File: server/src/controllers/activateAccount.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+const activateSchema = Joi.object({
+  active: Joi.boolean(),
+  admin: Joi.boolean(),
+});
+
+module.exports = (req, res, next) => {
+  const { admin, active } = req.body;
+  const { id } = req.params;
+
+  activateSchema
```
Comment (okjake) [inline server/src/controllers/activateAccount.js:15]: Since you have a good grasp of Promises I'd encourage you to also look at async / await syntax

File: server/src/controllers/booking.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
+const Boom = require('@hapi/boom');
+const nodemailer = require('nodemailer');
+const ical = require('ical-generator');
+const Moment = require('moment');
+const moment = require('../utils/moment-range');
+const { getBookingbydate } = require('../database/queries');
```
Comment (okjake) [inline server/src/controllers/booking.js:6]: Inconsistent capitalization here

File: server/src/controllers/booking.js [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    })
+    .catch(next);
+};
+
+const checkOverlap = (arrOfIntervals, interval) =>
+  arrOfIntervals.filter((existingInterval) =>
+    existingInterval.overlaps(interval)
+  );
+
+const bookingRoom = (req, res, next) => {
```
Comment (okjake) [inline server/src/controllers/booking.js:32]: This function is too big, I would look at the anonymous functions you are passing to `.then` everywhere, and see if you can pull them out into separate functions. It will make the flow easier to read and it's often easier to add unit tests for edge cases this way

File: server/src/controllers/deleteBooking.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    .then(({ rows }) => {
+      if (rows.length !== 0) {
+        return rows[0];
+      }
+      throw Boom.badRequest(
+        'the booking you are trying to delete does not exist'
+      );
+    })
+    .then((result) => {
+      if (role) {
```
Comment (okjake) [inline server/src/controllers/deleteBooking.js:18]: Is this right? It doesn't matter what the value of 'role' is?
